src/jupyterlabcontroller/storage/k8s.py

```diff
@@ -674,3 +682,430 @@ def _standard_metadata(self, name: str) -> V1ObjectMeta:
+                    "Error reading namespace",
+                    e,
+                    namespace=namespace,
+                ) from e
+        # Get all jobs
+        all_jobs = await self.batch_api.list_namespaced_job(namespace)
+        # Filter to those with the right label
+        users = [
+            x.metadata.labels.get("lsst.io/user")
+            for x in all_jobs.items
+            if x.metadata.labels.get("lsst.io/category", "") == "fileserver"
+        ]
+        # For each of these, check whether the fileserver is present
+        for user in users:
+            self._logger.debug(f"Checking user {user}")
+            good = await self.check_fileserver_present(
+                username=user, namespace=namespace
+            )
+            self._logger.debug(f"{user} fileserver good?  {good}")
+            if good:
+                observed_state[user] = True
+        return observed_state
+
+    async def check_fileserver_present(
+        self, username: str, namespace: str
+    ) -> bool:
+        """Our determination of whether a user has a fileserver is this:
+
+        We assume all fileserver objects are named <username>-fs, which we
+        can do, since we created them and that's the convention we chose.
+
+        A fileserver is working if:
+
+        1) it has a Job with at least one active Pod
+        2) it has a Service (properties not checked)
+        3) it has an Ingress (properties not checked)
+
+        We do not check the GafaelfawrIngress, because the Custom API is
+        clumsy, and the created Ingress is a good proxy.
+
+        If we find a broken fileserver, we delete all its objects.  In
+        steady-state operations, this will happen when a fileserver Pod has
+        seen no operations for its timeout period; the Pod will exit and
+        the reconciliation task will note that the Job has no active Pods
+        and take action to clean it up.
+        """
+        obj_name = f"{username}-fs"
+        self._logger.debug(f"Checking whether {username} has fileserver")
+        try:
+            self._logger.debug(f"Checking job for {username}")
+            job = await self.batch_api.read_namespaced_job(obj_name, namespace)
+        except ApiException as e:
+            self._logger.info(f"Job {obj_name} for {username} not found.")
+            if e.status == 404:
+                return False
+            raise KubernetesError.from_exception(
+                "Error reading job",
+                e,
+                namespace=namespace,
+                name=obj_name,
+            ) from e
+        if job.status is None or job.status.active < 1:
+            self._logger.info(
+                f"Job {obj_name} has no active pods; "
+                + f"terminating fileserver for {username}"
+            )
+            return False
+        try:
+            self._logger.debug(f"Checking service for {username}")
+            job = await self.api.read_namespaced_service(obj_name, namespace)
```

Comment:
Not sure you really need to do this; I don't see any reason why the `Service` object would get randomly deleted, unlike the `Pod` (which exits) and the `Ingress` (which has to be created by an operator). I suppose it can't hurt, but it may be wasted effort.

Reply:
Taken out.